File: game-server/src/main/java/net/dodian/uber/game/model/player/skills/mining/Mining.java
package net.dodian.uber.game.model.player.skills.mining;

import net.dodian.uber.game.model.entity.player.Client;
import net.dodian.uber.game.model.item.Equipment;
import net.dodian.uber.game.model.player.packets.incoming.ClickObject;
import net.dodian.uber.game.model.player.packets.outgoing.SendMessage;
import net.dodian.uber.game.model.player.skills.Skills;
import net.dodian.utilities.Misc;
import net.dodian.utilities.Utils;

public class Mining {

    public static int getMiningEmote(int item) {
        switch (item) {
            case 1275: //bronze
                return 624;
            case 1271: //iron
                return 628;
            case 1273: //steel
                return 629;
            case 1269: // mithril
                return 627;
            case 1267: // addy
                return 626;
            case 1265: // rune
                return 625;
            case 11920: //dragon
            case 20014: //3rd age
                return 7139;
        }
        return -1;
    }

    public static int findPick(Client player) {
        int Eaxe = -1, Iaxe = -1;
        int weapon = player.getEquipment()[Equipment.Slot.WEAPON.getId()];
        for (int i = 0; i < Utils.picks.length; i++) {
            if (Utils.picks[i] == weapon) {
                if (player.getLevel(Skills.MINING) >= Utils.pickReq[i])
                    Eaxe = i;
            }
            for (int ii = 0; ii < player.playerItems.length; ii++) {
                if (Utils.picks[i] == player.playerItems[ii] - 1) {
                    if (player.getLevel(Skills.MINING) >= Utils.pickReq[i]) {
                        Iaxe = i;
                    }
                }
            }
        }
        return Eaxe > Iaxe ? Eaxe : Iaxe > Eaxe ? Iaxe : -1;
    }

    public static long getMiningSpeed(Client player) {
        double pickBonus = Utils.pickBonus[player.minePick];
        double level = (double) player.getLevel(Skills.MINING) / 600;
        double random = (double) Misc.random(150) / 100;
        double bonus = 1 + pickBonus * random + level;
        double time = Utils.mineTimes[player.mineIndex] / bonus;
        return (long) time;
    }

    public static void Gemchance(Client player) {
        double[] chance = new double[]{7.03, 3.91, 3.91, 3.12};
        int[] gemId = new int[]{1623, 1621, 1619, 1617};
        int rolledChance = 0, gem = -1, roll = Misc.chance(10000);
        for (int i = 0; i < chance.length && gem == -1; i++) {
            rolledChance += (int) (chance[i] * 100);
            if (roll <= rolledChance) gem = gemId[i + 1];
            else if (i + 1 == chance.length) gem = gemId[0];
        }
        if (player.freeSlots() > 0) {
            player.addItem(gem, 1);
            player.send(new SendMessage("You managed to find a " + player.GetItemName(gem).toLowerCase() + " while mining."));
        }
    }

    public static void mining(int index, Client player) {
        if (!player.playerHasItem(-1)) {
            player.send(new SendMessage("There is not enough space in your inventory."));
            player.resetAction(true);
            return;
        } else
        if (index != 6) {
            player.send(new SendMessage("You mine some " + player.GetItemName(Utils.ore[index]).toLowerCase() + "."));
        }
        player.addItem(Utils.ore[index], 1);
        player.giveExperience(Utils.oreExp[index], Skills.MINING);
        player.triggerRandom(Utils.oreExp[index]);
        if (Misc.chance(30) == 1) {
            player.send(new SendMessage("You take a rest."));
            player.resetAction(true);
        } else if (Misc.chance(86) == 1) {//256 without glory
            Gemchance(player);
        }
    }
}

File: game-server/src/main/java/net/dodian/uber/game/model/player/skills/mining/Mining.kt
package net.dodian.uber.game.model.player.skills.mining

import net.dodian.uber.game.model.entity.player.Client
import net.dodian.uber.game.model.item.Equipment
import net.dodian.uber.game.model.player.skills.Skills
import net.dodian.utilities.Misc
import net.dodian.utilities.Utils
import java.util.*

    fun getMiningEmote(item: Int): Int {
        when (item) {
            1275 -> return 624
            1271 -> return 628
            1273 -> return 629
            1269 -> return 627
            1267 -> return 626
            1265 -> return 625
            11920, 20014 -> return 7139
        }
        return -1
    }
    fun findPick(player: Client): Int {
        var Eaxe = -1
        var Iaxe = -1
        val weapon = player.equipment[Equipment.Slot.WEAPON.id]
        for (i in Utils.picks.indices) {
            if (Utils.picks[i] == weapon) {
                if (player.getLevel(Skills.MINING) >= Utils.pickReq[i]) Eaxe = i
            }
            for (ii in player.playerItems.indices) {
                if (Utils.picks[i] == player.playerItems[ii] - 1) {
                    if (player.getLevel(Skills.MINING) >= Utils.pickReq[i]) {
                        Iaxe = i
                    }
                }
            }
        }
        return if (Eaxe > Iaxe) Eaxe else if (Iaxe > Eaxe) Iaxe else -1
    }

    fun getMiningSpeed(player: Client): Long {
        val pickBonus = Utils.pickBonus[player.minePick]
        val level = player.getLevel(Skills.MINING).toDouble() / 600
        val random = (50 + Misc.random(100)) / 100.0
        val bonus = 1 + pickBonus * random + level
        val time = Utils.mineTimes[player.mineIndex] / bonus
        return time.toLong()
    }

    fun gemChance(player: Client) {
        val chance = doubleArrayOf(7.03, 3.91, 3.91, 3.12)
        val gemId = intArrayOf(1623, 1621, 1619, 1617)
        var rolledChance = 0
        var gem = -1
        val roll = Misc.chance(10000)
        var i = 0
        while (i < chance.size && gem == -1) {
            rolledChance += (chance[i] * 100).toInt()
            if (roll <= rolledChance) gem = gemId[i + 1] else if (i + 1 == chance.size) gem = gemId[0]
            i++
        }
        if (player.freeSlots() > 0) {
            player.addItem(gem, 1)
            player.sendMessage("You managed to find a " + player.GetItemName(gem).lowercase(Locale.getDefault()) + " while mining.")
        }
    }

    fun mining(index: Int, player: Client) {
        doAction(player, Utils.ore[index], Utils.oreExp[index])
        if (Misc.chance(86) == 1) { //256 without glory
            gemChance(player)
        }
    }

    fun miningEss(p: Client) {
        if (p.getLevel(Skills.MINING) < 30) {
            doAction(p, 1436, 50)
        } else
            doAction(p, 7936, 50)
    }

    fun doAction(p: Client, i: Int, xp: Int) {
        if (!p.playerHasItem(-1)) {
            p.sendMessage("There is not enough space in your inventory.")
            p.resetAction(true)
            return
        }
        p.sendMessage("You mine some " + p.GetItemName(i).lowercase(Locale.getDefault()) + ".")
        p.addItem(i, 1)
        p.giveExperience(xp, Skills.MINING)
        p.triggerRandom(xp)
    }